src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs

```diff
@@ -386,6 +386,252 @@ IValue CopyAndUpdateUdt(TupleValue originalValue)
             return ResolvedExpressionKind.InvalidExpr;
         }
 
+        /// <summary>
+        /// Creates a string literal that contains the given values as interpolation arguments.
+        /// The given string is expected to consist of text interspaced with integer constants between curly brackets.
+        /// The content between non-escaped curly brackets is parsed as integer,
+        /// and replaced with the string representation of the interpolation argument at that index.
+        /// Registers the built string with the scope manager.
+        /// </summary>
+        internal static IValue CreateStringLiteral(GenerationContext sharedState, string str, params IValue[] exs)
+        {
+            static (int, int, int) FindNextExpression(string s, int start)
+            {
+                while (true)
+                {
+                    var i = s.IndexOf('{', start);
+                    if (i < 0)
+                    {
+                        return (-1, s.Length, -1);
+                    }
+                    else if ((i == start) || (s[i - 1] != '\\'))
+                    {
+                        var j = s.IndexOf('}', i + 1);
+                        if (j < 0)
+                        {
```

Comment:
For the future, my own subjective opinion, feel free to ignore. I can also be looking at the wrong links/topics.

The topic [`IndexOf(Char, Int32)`](https://docs.microsoft.com/en-us/dotnet/api/system.string.indexof?view=net-5.0#System_String_IndexOf_System_Char_System_Int32_) tells me that the call **returns -1** if the char is not found.
The code here checks the returned values like this: `if (i < 0)`, `if (j < 0)`.

If in the future the .NET API changes, and the function `IndexOf(Char, Int32)` starts returning (for certain cases) -2, -3, -4, and so on, then the checks `if (i < 0)`, `if (j < 0)` will start working in the unexpected way. The refactoring will be needed, that refactoring may be huge.

For the future code I recommend checking exactly against the value described/promised in the documentation:
`if (i == -1)`, `if (j == -1)`.
The .NET documentation writers are _free to make such a change_ (to start returning -2, -3), because the only thing that they promised is `-1` (not `< 0`).

Reply:
I changed it to -1. In general I agree that it is good to rely only on documented behavior and to not be "permissive" when writing code (trying to continue when dealing with an unexpected state/receiving unexpected input). In this specific case, anything negative will now result in an exception, since we are accessing an item at that index.